platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/Bucket.java

```diff
@@ -84,21 +85,57 @@
      */
     private ByteBuffer bucketBuffer;
 
-    private KeySerializer<K> keySerializer;
+    private final KeySerializer<K> keySerializer;
     private ByteBuffer reusableBuffer;
 
+    /**
+     * Bucket pool this bucket is managed by, optional. If not null, the bucket is
+     * released back to the pool on close.
+     */
+    private final ReusableBucketPool<K> bucketPool;
+
+    /**
+     * Create a new bucket with the default size.
+     *
+     * @param keySerializer The serializer responsible for converting keys to/from bytes
+     */
+    Bucket(final KeySerializer<K> keySerializer) {
+        this(keySerializer, null);
+    }
+
     /**
      * Create a new bucket with the default size.
      *
      * @param keySerializer The serializer responsible for converting keys to/from bytes
      */
-    Bucket(KeySerializer<K> keySerializer) {
-        setKeySerializer(keySerializer);
+    Bucket(final KeySerializer<K> keySerializer, final ReusableBucketPool<K> bucketPool) {
+        this.keySerializer = keySerializer;
+        this.keySerializationVersion = (int) keySerializer.getCurrentDataVersion();
         bucketBuffer = ByteBuffer.allocate(DEFAULT_BUCKET_BUFFER_SIZE);
         setSize(BUCKET_HEADER_SIZE);
         setBucketIndex(-1);
         reusableBuffer =
                 keySerializer.isVariableSize() ? ByteBuffer.allocate(keySerializer.getTypicalSerializedSize()) : null;
+        this.bucketPool = bucketPool;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void close() throws IOException {
+        if (bucketPool != null) {
+            bucketPool.releaseBucket(this);
+        }
```

Comment:
Do we need to close the reusableBuffer here?

Reply:
No. The buckets are reusable, that's the idea, so the buffer should be kept open.

Reply:
So closing a bucket still allows the same bucket to be reused?  Interesting.  Might be worth documenting that in JavaDoc, as it runs contrary to the general expectation.